docs/codeql/writing-codeql-queries/debugging-data-flow-queries-using-partial-flow.rst

```diff
@@ -0,0 +1,106 @@
+.. _debugging-data-flow-queries-using-partial-flow:
+
+Debugging data-flow queries using partial flow
+==============================================
+
+If a data-flow query unexpectedly produces no results, you can use partial flow to debug the problem.
+
+In CodeQL, you can use :ref:`data flow analysis <about-data-flow-analysis>` to compute the possible values that a variable can hold at various points in a program.
+A typical data-flow query looks like this:
+
+.. code-block:: ql
+
+::
+
+    class MyConfig extends TaintTracking::Configuration {
+      MyConfig() { this = "MyConfig" }
+
+      override predicate isSource(DataFlow::Node node) { node instanceof MySource }
+
+      override predicate isSink(DataFlow::Node node) { node instanceof MySink }
+    }
+
+    from MyConfig config, DataFlow::PathNode source, DataFlow::PathNode sink
+    where config.hasFlowPath(source, sink)
+    select sink.getNode(), source, sink, "Sink is reached from $@.", source.getNode(), "here"
+
+Or slightly simpler without path explanations:
+
+.. code-block:: ql
+
+    from MyConfig config, DataFlow::Node source, DataFlow::Node sink
+    where config.hasPath(source, sink)
+    select sink, "Sink is reached from $@.", source.getNode(), "here"
+
+If a data-flow query that you have written does not produce any results when you expect it to, there may be a problem with your query.
+You can try to debug the potential problem by following the steps described below. 
+
+Checking sources and sinks
+--------------------------
+
+As a first step, make sure that the source and sink definitions contain what you expect. If one of these is empty then there can never be any data flow. The easiest way to verify this is using quick evaluation in CodeQL for VS Code: Select the text ``node instanceof MySource``, right-click, and choose "CodeQL: Quick Evaluation". This will evaluate the highlighted text, which in this case means the set of sources.
```

Comment:
Is there a difference between selecting the text `node instanceof MySource` and selecting the text `isSource`?

Reply:
Not really. 

If you do a quick evaluation the `isSource` member-predicate, your output will contains 2 columns: `this` (which is the string that identifies the configuration), and `node` which has the relevant nodes that are sources. So you will end up with a bit more noise in the output :shrug: